### PR TITLE
Initial VSTS Guidance

### PR DESCRIPTION
**Pull request by markwilkie**

### CODE REVIEW

File: Documentation/VSTSGuidance.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+For now, everything should be in Dotnet-Internal and any code that is public should be on GitHub
+ 
+VSTS repos should:
+-  Be mirrored from GitHub, if not internal only.
+-  Internal-only projects should only be in the DotNet-Internal project with no github equivalent
+ 
+Naming conventions
+-  $(OrgName)-$(RepoName)-{Internal|Public*}   (we're not using 'trusted' anymore.  It's redundant)
+-  Again - *No plan to have public repos in VSTS at this time
+
```
Comment (chcosta) [inline Documentation/VSTSGuidance.md]: Add a note about folder structure for VSTS build definitions in repos that will create the build definition folder structure convention outlined.  https://github.com/Microsoft/vsts-agent/pull/1430/files#diff-0e4df20b2155d804a6518e8089072a96R29

```text
.vsts-pipelines
  builds/
    $(GitHubOrg)/
      $(GitHubRepoName)/
        $(OrgName)-$(RepoName)-{Internal|Public}.yml
```

Comment (alexperovich) [reply Documentation/VSTSGuidance.md]: Do we really need the Internal/Public in the build names? The builds are going to be in completely different projects in VSTS.

File: Documentation/VSTSGuidance.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# VSTS Guidance
+
+## Projects
+There are two projects for use.  They are:
+-  DotNet-Public
+    -  Used for CI
+    -  For build definitions only  (no source code - that's on github)
+    -  Build definitions are allowed to pull source directly from GitHub
```
Comment (weshaggard) [inline Documentation/VSTSGuidance.md]: Will we have any repo's that produce official builds from github sources?

File: Documentation/VSTSGuidance.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+Being figured out...
+
+## Permissions
+
+Being figured out...
+
+## Build Definitions
+
+Folder names for github repos: 
+- $(GitHubOrg)/$(GitHubRepoName)/*.def
```
Comment (weshaggard) [inline Documentation/VSTSGuidance.md]: This looks like it will be the vsts build definition name, correct?

File: Documentation/VSTSGuidance.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+## Teams
+
+Being figured out...
+
+## Permissions
+
+Being figured out...
+
+## Build Definitions
```
Comment (weshaggard) [inline Documentation/VSTSGuidance.md]: I think we should also define where the build definition file in the repo's should go. Currently it looks like we have a ".vsts-ci.yml" file in the root of the repo's but I suggest we put it somewhere under the "eng" directory so we can try and keep the root directory as clean as possible.

Comment (chcosta) [reply Documentation/VSTSGuidance.md]: I don't know that they'll let us put it in the eng folder specifically. You can comment on https://github.com/Microsoft/vsts-agent/pull/1430/files#diff-0e4df20b2155d804a6518e8089072a96R29 if you feel strongly that this should be supported.

File: Documentation/VSTSGuidance.md (hunk 2 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: No.  Being secure (provenance specifically) says that we must only use resources from end points we (msft) own for official builds.

Comment (weshaggard) [reply Documentation/VSTSGuidance.md]: If that is the case why would we pull directly from github? Also I'm a little confused about your comment below "No plan to have public repos in VSTS at this time"

Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: Another way of saying it is that the public side of VSTS is only used for CI - and CI would indeed pull directly from gh.  However, for our official builds, we'd use the internal project and "mirror" gh into the repo on VSTS, then build from there.

Once I've managed to be more clear, I'll update the doc to improve.  Let me know if we're there yet...

File: Documentation/VSTSGuidance.md (hunk 1 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: So this section of the doc is about repos, not build definitions.  And yes, knowing if a repo is internal/public is good to know.

File: Documentation/VSTSGuidance.md (hunk 3 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: The intent is to denote the path.  The name guidance is further expanded slightly later in the doc.  Did I understand your question correct?

File: Documentation/VSTSGuidance.md (hunk 1 above)
Comment (chcosta) [reply Documentation/VSTSGuidance.md]: Aren't the build definitions defined in the repo and then expressed via build definitions in VSTS?

Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: Yes, and to your point, I don't think this doc addresses both the definition itself and the yml file.  @chcosta, any chance you could share exactly what are options are?  I know we're restricted in some ways....

File: Documentation/VSTSGuidance.md (hunk 2 above)
Comment (weshaggard) [reply Documentation/VSTSGuidance.md]: Getting closer...

Maybe I'm wrong but I thought there is going to be only one build definition for CI and official with perhaps an additional property being set in one or the other. In that case it would seem the build definition needs to be able to pull from either VSTS or github. Even if we do have multiple build definitions none for official and one for CI I still think we need to support running CI on either side, github or vsts PRs (perhaps that is yet another build definition but I hope not).

File: Documentation/VSTSGuidance.md (hunk 3 above)
Comment (weshaggard) [reply Documentation/VSTSGuidance.md]: The label says folder names for github repos but I'm unsure where these folder go. Are those folders in VSTS?

File: Documentation/VSTSGuidance.md (hunk 2 above)
Comment (chcosta) [reply Documentation/VSTSGuidance.md]: My understanding is that VSTS will set a property (env variable) which can be used at build time to determine if you are running in official build or CI context.  The intent there is to share the same definition with conditioned steps.

File: Documentation/VSTSGuidance.md (hunk 1 above)
Comment (chcosta) [reply Documentation/VSTSGuidance.md]: The current design restricts the repo folder structure to something like this...

```text
.vsts-pipelines/ 
  builds/
   [optional subfolders]/
     [build name].yml 
  releases/
   [optional subfolders]/
    [build name].yml 
```

That gets expressed in VSTS as "build" or "release" definitions with the same "optional subfolder" structure and "build name" named definition.

File: Documentation/VSTSGuidance.md (hunk 2 above)
Comment (mmitche) [reply Documentation/VSTSGuidance.md]: We will have multiple build definitions, effectively a mirrored set in the public and internal.  We will have one set of YAML which can apply to both.  This will allow for CI (PR testing) in both internal and OSS venues, and well as official build production on the internal side.  Theoretically you could have non-signed per-commit builds in OSS, though they may be less useful.

File: Documentation/VSTSGuidance.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+## Teams
+
+In this context, teams pretty much only affect which level the kanban (or whatever) boards are at.  While permissions can be set at the team level, we're choosing not to do so.
+
+Teams:
+-  $(GitHubOrg)
+
+## Permissions
+To keep things as simple (manageable) as possible, we're going to manage permissions coarsely at the project level - pointing directly to existing AD security groups managed in idweb.  **We should not be managing permission outside of this method**
```
Comment (mmitche) [inline Documentation/VSTSGuidance.md]: If permissions are relatively course, then we should be able to lock down adminstrative permissions enough to avoid issues.

File: Documentation/VSTSGuidance.md (hunk 3 above)
Comment (mmitche) [reply Documentation/VSTSGuidance.md]: @weshaggard These folders are in VSTS.  There is a folder structure for build definitions.  In DevDiv's instance, we haven't used this much, and instead put everything at the top level.

@markwilkie There is a coming feature that will allow us to simply point VSTS at a folder of a repo and have it create build definitions for every yaml file there.  This should make management easy.

File: Documentation/VSTSGuidance.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    - code-coverage
+    - slow-tests
+    - fast-tests
+    - internal-tools (TBD -- Nate and Matt to investigate more)
+    - official
+  - Suffix:
+    - internal = dotnet-internal builds
+    - public = dotnet-public builds
+ 
+TODO: figure out how to control definition names automatically created by VSTS
```
Comment (mmitche) [inline Documentation/VSTSGuidance.md]: @markwilkie I think this may come automatically with their folder-watching.  May want to follow up with VSTS though.

File: Documentation/VSTSGuidance.md (hunk 1 above)
Comment (mmitche) [reply Documentation/VSTSGuidance.md]: @alexperovich It's not always obvious what you're looking at in the UI, since the public/internal shows up only in two places:
- Upper left hand corner
- URL

I wanted to avoid confusion.

File: Documentation/VSTSGuidance.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+## Projects
+There are two projects for use.  They are:
+-  DotNet-Public
+    -  Used for CI
+    -  For build definitions only  (no source code - that's on github)
+    -  Build definitions are allowed to pull source directly from GitHub
+-  DotNet-Internal
+    -  Build definitions are only allowed to pull source from internal repos
+ 
+TODO: figure out the deal with default repo
```
Comment (vtbassmatt) [inline Documentation/VSTSGuidance.md]: I'd recommend putting a readme.md into it, which basically says "This is where the builds go. You probably want to check us out over on GitHub -->" with a link to whatever makes sense on GitHub. readme.md in the default repo will get automatically loaded on the project homepage (https://dotnet.visualstudio.com/DotNet-Public/).

File: Documentation/VSTSGuidance.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+In this context, teams pretty much only affect which level the kanban (or whatever) boards are at.  While permissions can be set at the team level, we're choosing not to do so.
+
+Teams:
+-  $(GitHubOrg)
+
+## Permissions
+To keep things as simple (manageable) as possible, we're going to manage permissions coarsely at the project level - pointing directly to existing AD security groups managed in idweb.  **We should not be managing permission outside of this method**
+
+-  Permissions will point to existing security groups in AD which are managed in idweb.  This admin is done at the **project** level.  ([vsts link](https://dotnet.visualstudio.com/DotNet-Internal/_admin/_securit))
```
Comment (vtbassmatt) [inline Documentation/VSTSGuidance.md]: Typo in link to project-level security, missing the 'y' at the end. Also, nitpick - can we spell the product acronym "VSTS" in capital letters?

File: Documentation/VSTSGuidance.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+-  The bulk of folks will be in the 'contributers' group, with special additions for other groups (like admin)
+-  There are VSTS permission groups that can be set out side of the project. ([vsts link](https://dotnet.visualstudio.com/_admin/_security))   **We're not going to use those**
```
Comment (vtbassmatt) [inline Documentation/VSTSGuidance.md]: "There are VSTS groups than can be assigned permissions outside the project context." would be a little more clear (to me, anyhow).

File: Documentation/VSTSGuidance.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+- kebab-case, No spaces, use dashes
+- Pattern: $(RepoName) - $Scenario - $Suffix
+  - Scenario: (optional)
+    - code-coverage
+    - slow-tests
+    - fast-tests
+    - internal-tools (TBD -- Nate and Matt to investigate more)
+    - official
+  - Suffix:
+    - internal = dotnet-internal builds
```
Comment (ChadNedzlek) [inline Documentation/VSTSGuidance.md]: Is this "internal" the same "internal" at the project level? Is that not redundant?

File: Documentation/VSTSGuidance.md (hunk 1 above)
Comment (ChadNedzlek) [reply Documentation/VSTSGuidance.md]: If we thought "Trusted" is redundant, I'm not 100% sure why "Internal" isn't equally redundant... they seem to be different words for the same thing.  Or are there cases where "Internal" and "Trusted" are non-synonymous?

Having "internal" all over the place seems like a lot of repeating the same data... doesn't this doc mean the path to a build is going to be:
/Internal/DotNet-HelixMachines-Internal/builds/DotNet-HelixMachines-Internal/DoetNet-HelixMachines-CreateImages-Internal.yml?

That's a lot of repeating the same thing quite a bit... especially since when mirrored to my computer, I'm not going to know what the "right" name for these files is, because I'm certainly going to override the clone so that I don't have gigantic folder names all over my computer. Is there no way to know the context without having to repeat it quite so often? Like perhaps a request for VSTS to make "important" stuff more prominent in the UI?

Comment (ChadNedzlek) [reply Documentation/VSTSGuidance.md]: That also means that if you want to mirror/move the builds to somewhere else, the repository name is repeated inside the repository itself, which might be pretty confusing to see in the history, and potentially be information we might not want in the files themselves.

Comment (ChadNedzlek) [reply Documentation/VSTSGuidance.md]: It's also possible that we will promote "internal" things to public once they shape up (I know there are always rumors of this for Helix)... It feels odd for something inside the repository to refer to the name (or location) of the repository itself.  Sort of like if my folders on my computer were all a/a-b/a-b-c/a-b-c-d... Makes it hard to move and restructure things cleanly later when every node in the tree knows about all the other nodes in the tree.

File: Documentation/VSTSGuidance.md (hunk 3 above)
Comment (shawnro) [inline Documentation/VSTSGuidance.md]: I find it helpful to see a specific example anywhere variable patterns are used to help alleviate confusion.

File: Documentation/VSTSGuidance.md (hunk 2 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: Updated doc to make more clear

File: Documentation/VSTSGuidance.md (hunk 3 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: I've referenced the coming feature in the yml folder section

File: Documentation/VSTSGuidance.md (hunk 5 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: Correct - this is the thinking

File: Documentation/VSTSGuidance.md (hunk 2 above)
Comment (chcosta) [reply Documentation/VSTSGuidance.md]: > We will have one set of YAML which can apply to both.

@mmitche , how are you imaging that this will be managed?  ie, the build definitions are populated from the checked in yaml.  How do you get a build definition to populate to Dotnet-Internal and to Dotnet-Public using the same yaml?  Seems like *maybe* templates help here, but the pieces don't all align in my head to make this possible.  The alternative is to let the Dotnet-Public definitions be auto-created and manually create the Dotnet-Internal definitions (and specify a VSTS source).

Comment (chcosta) [reply Documentation/VSTSGuidance.md]: That's the issue I'm thinking of.  It'd be nice to avoid having to manually create build defs (though they should just be very simple build defs which include a source (VSTS or GH), and point to the yaml file.  Seems like we could at least get away with the public build defs getting auto-created though.

File: Documentation/VSTSGuidance.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
+# VSTS Guidance
+
+## Projects
+There are two projects for use.  They are:
```
Comment (JohnTortugo) [inline Documentation/VSTSGuidance.md]: A few question:
- Are these project / repo, specifically to store build definitions?
- What are the “internal repos” ? Are these other .NET Core repos or just forks of the Github ones?
- What is the deal with the “internal repos”? Is it something related to signing the packages? Security fixes?
- Do we we have the URL for these repo/projects?

File: Documentation/VSTSGuidance.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+-  Permissions will point to existing security groups in AD which are managed in idweb.  This admin is done at the **project** level.  ([VSTS link](https://dotnet.visualstudio.com/DotNet-Internal/_admin/_security))
+-  The bulk of folks will be in the 'contributers' group, with special additions for other groups (like admin)
+-  There are VSTS permission groups that can be set out side of the project context. ([VSTS link](https://dotnet.visualstudio.com/_admin/_security))   **We're not going to use those**
+-  It is also possible to set permissions at the team.  **We're not going to do that**
+
+## Build Definitions
+
+### Folder names for github repos: 
+- $(GitHubOrg)/$(GitHubRepoName)/*.def
```
Comment (JohnTortugo) [inline Documentation/VSTSGuidance.md]: - Isn’t there a /FolderName/*.def missing here? It seems that the above schema describe something like: dotnet/corefx/*.def and thus *.def would be in the root of the repo..?

File: Documentation/VSTSGuidance.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+ 
+### Folders for VSTS repos:
```
Comment (JohnTortugo) [inline Documentation/VSTSGuidance.md]: - So, some build definitions are stored in Github and some in VSTS, right? Is that related to security?
- Can you give me an example of these folder structure? Would this be an correct folder name: “dotnet/DotNet-Public/*.def” ?

File: Documentation/VSTSGuidance.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+### Build definition file name convention:
+- kebab-case, No spaces, use dashes
+- Pattern: $(RepoName) - $Scenario - $Suffix
+  - Scenario: (optional)
+    - code-coverage
+    - slow-tests
+    - fast-tests
+    - internal-tools (TBD -- Nate and Matt to investigate more)
+    - official
+  - Suffix:
```
Comment (JohnTortugo) [inline Documentation/VSTSGuidance.md]: - Why is the suffix needed? Won’t internal builds be stored in VSTS and public ones in Github?

File: Documentation/VSTSGuidance.md (hunk 11 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: > Are these project / repo, specifically to store build definitions?
Not really, it's just that the build defs are stored in the repo
> What are the “internal repos” ? Are these other .NET Core repos or just forks of the Github ones?
"internal" is the mirror of our external github repos for the most part.  There will be some internal only, but those are the exception.  
> What is the deal with the “internal repos”? Is it something related to signing the packages? Security fixes?
For the most part, internal is just a one way mirror from github.  However, for things like MSRC and private features, yes, it'll be internal only.
> Do we we have the URL for these repo/projects?
Yes, I'll update the doc with these.

File: Documentation/VSTSGuidance.md (hunk 12 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: There's a new feature coming from VSTS which will allow for non root defs

File: Documentation/VSTSGuidance.md (hunk 13 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: >So, some build definitions are stored in Github and some in VSTS, right? Is that related to security?

This line is trying to say that for those repos in github, here's where the build def lives.  I'll try and clarify in the doc...

> Can you give me an example of these folder structure? Would this be an correct folder name: “dotnet/DotNet-Public/*.def” ?

I've updated the example below, but it's something like dotnet/arcade/DotNet-Arcade-Internal

File: Documentation/VSTSGuidance.md (hunk 14 above)
Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: The definitions get jumbled together and it's tough to know what you're looking at.  Public defs are for CI, and internal is for our official builds.  I'll try and clarify the doc.

File: Documentation/VSTSGuidance.md (hunk 13 above)
Comment (chcosta) [reply Documentation/VSTSGuidance.md]: Given, https://github.com/dotnet/arcade/issues/148, do we want to establish a convention around casing?  ie, dotnet/arcade/dotnet-arcade-internal?

Comment (markwilkie) [reply Documentation/VSTSGuidance.md]: ah - good point.  I'll update the doc.

Comment (alexperovich) [reply Documentation/VSTSGuidance.md]: why does the yaml file for dotnet/arcade (which lives in dotnet/arcade) need to have dotnet/arcade in its file name/path twice? We are repeating the repository name a ton of times.

File: Documentation/VSTSGuidance.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+## Projects
+There are two projects for use.  They are:
+-  DotNet-Public (https://dotnet.visualstudio.com/DotNet-Public)
+    -  Used for CI
+    -  For build definitions only  (no source code - that's on github)
+    -  Build definitions are allowed to pull source directly from GitHub
+-  DotNet-Internal  (https://dotnet.visualstudio.com/DotNet-Internal)
+    -  Build definitions are only allowed to pull source from internal repos
+    -  Public github repos should be mirror here for official msft builds
```
[the remark this answers is not in the document]
Comment (mmitche) [reply Documentation/VSTSGuidance.md]: Even if you have to duplicate, you don't have to duplicate logic.  Logic lives in source, and the build definition just becomes "look at this source"
